### PR TITLE
Add test to receive D2C messages with retain flag

### LINKED ISSUE #8: Client can send retained messages
Client send D2C message with RETAIN flag set to 1
Verify received
Verify x-opt-retain application property in the message

### PR DESCRIPTION
**Pull request by Ioana37**
Closes #8

### CODE REVIEW

File: Device/Device.cs [hunk 1]
```diff
@@ -68,36 +60,44 @@ public async Task ConnectDevice()
...
 
-        public async Task SendCloudToDeviceMessageAsync(string payload)
-        {
-            var topicD2C = $"devices/{deviceId}/messages/events/";
-
-            Console.WriteLine($"Topic:{topicD2C} Payload:{payload}");
-
+        public MqttApplicationMessage ConstructMessage(string topic, string payload, bool retainFlag = false)
+        {
+            /*var payloadJObject = new JObject
```
Comment (EliiseS) [inline Device/Device.cs]: Remove commented out code

File: Device/Device.cs [hunk 2]
```diff
@@ -68,36 +60,44 @@ public async Task ConnectDevice()
...
-                .Build();
+                .Build();
+
+            return message;
+        }
+
+
+        public async Task SendDeviceToCloudMessageAsync(string payload, bool retainFlag = false)
+        {
+            var topicD2C = $"devices/{deviceId}/messages/events/$.ct=application%2Fjson&$.ce=utf-8";
```
Comment (EliiseS) [inline Device/Device.cs:70]: Move variable topicD2C to ConstructMessage method since it's only used there

File: IotHubConsumer/Program.cs [hunk 3]
```diff
@@ -9,13 +10,28 @@ class Program
...
-
             var configuration = Configuration.BuildConfiguration();
 
-            Console.WriteLine("Enter message payload for C2D");
-            var input = Console.ReadLine();
-            var consumer = new Consumer(configuration);
-            consumer.SendCloudToDeviceMessageAsync(input).Wait();
+            //Console.WriteLine("Enter message payload for C2D");
+            // var input = Console.ReadLine();
+            var consumer = new SenderConsumer(configuration);
```
Comment (EliiseS) [inline IotHubConsumer/Program.cs]: You've changed this function in a way that no messages are sent from here anymore. Perhaps we should move the Sender consumer into it's own project with it's own program, since we can't run both of them from here anymore? And then not call it a sender consumer anymore :)

File: Testing/C2DTests.cs [hunk 4]
```diff
@@ -41,12 +44,34 @@ public async Task SendC2DMessages()
...
+
+            // Act
+            // send D2C with retain flag set to true
+            await device.SendDeviceToCloudMessageAsync(payload, retainFlag);
+
+            using var cancellationSource = new CancellationTokenSource();
+
+            // Assert - verify received + mqtt-retain set to true
+            cancellationSource.CancelAfter(3000);
+            var recConsumer = new ReceiverConsumer(configuration);
```
Comment (EliiseS) [inline Testing/C2DTests.cs]: I'd move this declaration out to the constructor with the others

File: Testing/C2DTests.cs [hunk 5]
```diff
@@ -41,12 +44,34 @@ public async Task SendC2DMessages()
...
+            await device.ConnectDevice();
+
+            var retainFlag = true;
+            var payload = Guid.NewGuid().ToString();
+
+            // Act
+            // send D2C with retain flag set to true
+            await device.SendDeviceToCloudMessageAsync(payload, retainFlag);
+
+            using var cancellationSource = new CancellationTokenSource();
```
Comment (EliiseS) [inline Testing/C2DTests.cs]: Is the using statement needed here?

File: Testing/C2DTests.cs [hunk 6]
```diff
@@ -41,12 +44,34 @@ public async Task SendC2DMessages()
...
+        public async Task SendD2CWithRetainFlagTrue()
+        {
+            // Arrange
+            await device.ConnectDevice();
+
+            var retainFlag = true;
+            var payload = Guid.NewGuid().ToString();
+
+            // Act
+            // send D2C with retain flag set to true
```
Comment (EliiseS) [inline Testing/C2DTests.cs]: Nitpick: I'd remove this comment due it being redundant

File: IotHubConsumer/ReceiverConsumer.cs [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    {
+        private readonly IConfiguration configuration;
+        public HashSet<TestElement> MessagesWithRetainSet { get; set; } = new HashSet<TestElement>();
+
+        public ReceiverConsumer(IConfiguration configuration)
+        {
+            this.configuration = configuration;
+        }
+
+        // Asynchronously create a PartitionReceiver for a partition and then start
```
Comment (EliiseS) [inline IotHubConsumer/ReceiverConsumer.cs]: I'd remove the comments in this file

File: IotHubConsumer/ReceiverConsumer.cs [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+                await foreach (var partitionEvent in consumer.ReadEventsAsync(cancellationToken))
+                {
+                    Console.WriteLine("Message received on partition {0}:", partitionEvent.Partition.PartitionId);
+
+                    var data = Encoding.UTF8.GetString(partitionEvent.Data.Body.ToArray());
+
+                    if (partitionEvent.Data.Properties.ContainsKey("mqtt-retain")) //TODO: can remove this if and add all messages for other unit tests
+                    {
+                        MessagesWithRetainSet.Add(new TestElement { Payload = data, RetainFlag = partitionEvent.Data.Properties["mqtt-retain"].ToString() });
```
Comment (EliiseS) [inline IotHubConsumer/ReceiverConsumer.cs]: I'd add the partitionEvent directly into the list, instead of creating a new object for it

File: IotHubConsumer/ReceiverConsumer.cs [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+                //   https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/eventhub/Azure.Messaging.EventHubs.Processor/README.md
+
+
+                await foreach (var partitionEvent in consumer.ReadEventsAsync(cancellationToken))
+                {
+                    Console.WriteLine("Message received on partition {0}:", partitionEvent.Partition.PartitionId);
+
+                    var data = Encoding.UTF8.GetString(partitionEvent.Data.Body.ToArray());
+
+                    if (partitionEvent.Data.Properties.ContainsKey("mqtt-retain")) //TODO: can remove this if and add all messages for other unit tests
```
Comment (EliiseS) [inline IotHubConsumer/ReceiverConsumer.cs]: This blocks other tests from being run without removing/changing it, I'd remove it

File: IotHubConsumer/ReceiverConsumer.cs [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+using System.Threading.Tasks;
+using Azure.Messaging.EventHubs.Consumer;
+using Microsoft.Extensions.Configuration;
+
+namespace IotHubConsumer
+{
+    public class ReceiverConsumer
+    {
+        private readonly IConfiguration configuration;
+        public HashSet<TestElement> MessagesWithRetainSet { get; set; } = new HashSet<TestElement>();
```
Comment (EliiseS) [inline IotHubConsumer/ReceiverConsumer.cs]: Could you not return this list ReceiveMessagesFromDeviceAsync method, instead of making it a public variable?

Comment (EliiseS) [inline IotHubConsumer/ReceiverConsumer.cs]: Why did decide to use a hashset? :)

File: Device/Device.cs [hunk 11]
```diff
@@ -68,36 +60,44 @@ public async Task ConnectDevice()
...
 
-        public async Task SendEventsAsync()
+        public async Task SendD2CMessagesInALoopAsync() // TODO: delete if no longer used
         {
+            var payload = 0;
             while (true)
             {
-                var payloadJObject = new JObject();
-
-                payloadJObject.Add("OfficeTemperature", "22." + DateTime.UtcNow.Millisecond.ToString());
```
Comment (EliiseS) [inline Device/Device.cs:94]: Why did you remove the payload creation here?

**Review by EliiseS — CHANGES_REQUESTED** [review]
I'd also change the PR name to be a bit more readable :)

File: Device/Device.cs (hunk 11 above)
Comment (Ioana37) [reply Device/Device.cs:94]: To be able to control the content of the payload from only what I pass as a parameter. And also to be able to verify the value in the unit test without serialising/deserialising.

File: IotHubConsumer/ReceiverConsumer.cs (hunk 10 above)
Comment (Ioana37) [reply IotHubConsumer/ReceiverConsumer.cs]: It started as a hashset of strings for which I wanted to have retrieval in O(1) when searching for a value (PTSD after coding interviews 🙄). Depending on whether we decide to keep the TestElement class or not, I could also create a Dictionary with the guid of the message as key, but that might complicate adding the other properties as value.
Which other data structures did you have in mind?

File: IotHubConsumer/Program.cs (hunk 3 above)
Comment (Ioana37) [reply IotHubConsumer/Program.cs]: Good point!  Will think how I can change it.

File: IotHubConsumer/ReceiverConsumer.cs (hunk 10 above)
Comment (Ioana37) [reply IotHubConsumer/ReceiverConsumer.cs]: I did not want to modify the behaviour of the method, in case we use in other places except for the unit test.

Comment (EliiseS) [reply IotHubConsumer/ReceiverConsumer.cs]: I think in that case, it would make sense to try to make two public functions, one that returns the list when after cancelation and a second without, with a shared private method that contains the shared logic.

I feel the current implementation isn't very intuitive to use.

File: Device/Device.cs (hunk 11 above)
Comment (EliiseS) [reply Device/Device.cs:94]: If you want to do that why not just use the send message a single message function? I created this one to see constant stream of semi random messages sent to Iot hub.

File: IotHubConsumer/ReceiverConsumer.cs (hunk 10 above)
Comment (EliiseS) [reply IotHubConsumer/ReceiverConsumer.cs]: I'm pretty simple minded when it comes to data types and would go with icollectible of type partitionEvent. I made the suggestion of partitionEvent, because you're currently sending two parameters (plus more to come for other tests) from the event anyway, so why not just send the entire event?

I asked about the hash because I was curious :)

File: Device/Device.cs (hunk 2 above)
Comment (syedhassaanahmed) [reply Device/Device.cs:70]: Not sure about moving `topicD2C` to `ConstructMessage()`. Passing it from outside allows us to construct a message that can be sent to a different topic (even though we're currently not doing it).

File: Device/Program.cs [hunk 12]
```diff
@@ -14,7 +14,9 @@ public static async Task<int> Main(string[] args)
         {
             var configuration = Configuration.BuildConfiguration();
             var device = new Device(configuration);
```
Comment (syedhassaanahmed) [inline Device/Program.cs]: It should have been addressed in earlier PR :) but we should avoid passing configuration directly into the Device class. Device class is only concerned with IoT Hub connection string. Problem with passing entire config is, the Device now has to know about the configuration data structure and it also has access to other configurations even though the Device doesn't necessarily need them.

File: IotHubConsumer/ReceiverConsumer.cs (hunk 10 above)
Comment (syedhassaanahmed) [reply IotHubConsumer/ReceiverConsumer.cs:14]: @Ioana37 Did you use `TestElement` to avoid deserialization which would happen in case of `partitionEvent`?

File: Testing/C2DTests.cs (hunk 4 above)
Comment (syedhassaanahmed) [reply Testing/C2DTests.cs]: @EliiseS Did you mean "out to the Setup() method with the others"?

Comment (EliiseS) [reply Testing/C2DTests.cs]: Yes, exactly 😄 Setup, constructor, pish, posh

File: Device/Device.cs (hunk 2 above)
Comment (EliiseS) [reply Device/Device.cs:70]: Then perhaps, we should add it as a parameter for this method as well? Right now it's in the middle of those two worlds, which I don't think makes sense

File: IotHubConsumer/ReceiverConsumer.cs (hunk 8 above)
Comment (Ioana37) [reply IotHubConsumer/ReceiverConsumer.cs]: If we add the eventPartition object to the list, we would have to deserialise its data content when we do the lookup in the tests, which would make it more complex. If we end up using the message list in other tests, we would have the content already stored as a string.

File: IotHubConsumer/ReceiverConsumer.cs (hunk 10 above)
Comment (Ioana37) [reply IotHubConsumer/ReceiverConsumer.cs]: Tried to do some refactoring to increase readability. Once we add the offset, we should get rid of that list without changing the method.

File: Device/Device.cs [hunk 13]
```diff
@@ -128,17 +97,18 @@ private async void Disconnected(MqttClientDisconnectedEventArgs e, IMqttClientOp
             {
                 Console.WriteLine("### RECONNECTING FAILED ###");
             }
+
             Console.WriteLine("Reconnected");
         }
 
-        private static string GenerateSasToken(string resourceUri, string key, int expiryInSeconds = 36000)
```
Comment (EliiseS) [inline Device/Device.cs]: Why remove static?

File: Device/Program.cs [hunk 14]
```diff
@@ -10,12 +7,18 @@ namespace MessageSample
...
-            await device.RunSampleAsync(ApplicationMessageReceived);
-            return 0;
+            var configuration = Configuration.BuildConfiguration();
+            var iotHubDeviceConnectionString = configuration["IotHubDeviceConnectionString"];
+            var device = new Device(iotHubDeviceConnectionString);
+
+            await device.ConnectDevice();
+            await device.SubscribeToEventAsync(ApplicationMessageReceived);
+
+            Console.WriteLine("Enter message payload for D2C");
```
Comment (EliiseS) [inline Device/Program.cs]: Could we at least put this in a while loop with a command to exit? I'd like to keep the functionality of continuously sending messages

File: IotHubConsumer/Receiver.cs [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+
+        public async Task ReceiveMessagesFromDeviceAsync(CancellationToken cancellationToken)
+        {
+            await using var consumer = new EventHubConsumerClient(EventHubConsumerClient.DefaultConsumerGroupName, eventHubCompatibleEndpoint, eventHubName);
+
+            Console.WriteLine("Listening for messages on all partitions");
+
+            try
+            {
+                // Begin reading events for all partitions, starting with the first event in each partition and waiting indefinitely for
```
Comment (EliiseS) [inline IotHubConsumer/Receiver.cs]: I'd still remove this comment :)

File: IotHubConsumer/Receiver.cs [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+            }
+            catch (TaskCanceledException)
+            {
+                // This is expected when the token is signaled; it should not be considered an
+                // error in this scenario.
+            }
+        }
+    }
+
+    public class D2CMessage
```
Comment (EliiseS) [inline IotHubConsumer/Receiver.cs]: I'd move this class into it's own file

File: IotHubConsumer/Receiver.cs [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+                    var data = Encoding.UTF8.GetString(partitionEvent.Data.Body.ToArray());
+                    Console.WriteLine("\t{0}:", data);
+
+                    var retainFlag = "false";
+                    if (partitionEvent.Data.Properties.ContainsKey("mqtt-retain"))
+                    {
+                        retainFlag = partitionEvent.Data.Properties["mqtt-retain"].ToString();
+                    }
+
+                    ReceivedMessages.Add(new D2CMessage { Payload = data, RetainFlag = retainFlag });
```
Comment (EliiseS) [inline IotHubConsumer/Receiver.cs]: Why not add all the properties instead of the just retain flag?

File: IotHubConsumer/Receiver.cs [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+                //
+                // The "ReadEventsAsync" method on the consumer is a good starting point for consuming events for prototypes
+                // and samples.  For real-world production scenarios, it is strongly recommended that you consider using the
+                // "EventProcessorClient" from the "Azure.Messaging.EventHubs.Processor" package.
+                //
+                // More information on the "EventProcessorClient" and its benefits can be found here:
+                //   https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/eventhub/Azure.Messaging.EventHubs.Processor/README.md
+
+
+                await foreach (var partitionEvent in consumer.ReadEventsAsync(cancellationToken))
```
Comment (EliiseS) [inline IotHubConsumer/Receiver.cs:35]: Can you add a offset?

File: Device/Device.cs (hunk 13 above)
Comment (Ioana37) [reply Device/Device.cs]: I thought it is no longer in a static class and it is a private method whose parameters depend on properties from this class.

File: IotHubConsumer/Program.cs [hunk 19]
```diff
@@ -1,21 +1,51 @@
...
+        public static void SendCloudToDeviceMessage(IConfiguration configuration)
+        {
             Console.WriteLine("Enter message payload for C2D");
             var input = Console.ReadLine();
-            var consumer = new Consumer(configuration);
+
+            var iotHubConnectionString = configuration["IotHubConnectionString"];
+            var deviceId = configuration["DeviceId"];
+            var consumer = new Sender(iotHubConnectionString, deviceId);
             consumer.SendCloudToDeviceMessageAsync(input).Wait();
```
Comment (EliiseS) [inline IotHubConsumer/Program.cs]: Use async keyword instead of wait (for all instances)

File: Device/Device.cs (hunk 2 above)
Comment (syedhassaanahmed) [reply Device/Device.cs:70]: @EliiseS I don't disagree with adding `topicD2C` as a param to `SendDeviceToCloudMessageAsync()`. The only caveat would be to then parse `deviceId` from the connection string twice - once inside this class constructor and another in the caller of `SendDeviceToCloudMessageAsync()`. May I suggest we postpone this until we see if other tests actually need to change the topic name at all? If it will always be the same, we could move it inside `ConstructMessage()` like you originally suggested.

File: IotHubConsumer/Receiver.cs (hunk 17 above)
Comment (syedhassaanahmed) [reply IotHubConsumer/Receiver.cs]: All properties are not needed yet for the tests, lets add properties when they're needed.

File: Device/Device.cs (hunk 2 above)
Comment (EliiseS) [reply Device/Device.cs:70]: Sure, sure :)

File: IotHubConsumer/Receiver.cs (hunk 18 above)
Comment (EliiseS) [reply IotHubConsumer/Receiver.cs:35]: Adding a offset later :)

**Review by EliiseS — APPROVED** [review]
